README.md
# Flood integration for Home Assistant

This a _custom component_ for [Home Assistant](https://www.home-assistant.io/).
The `flood` integration allows you to get information and control your [Flood WebUI torrent client](https://github.com/jesec/flood), based on last @jesec version.

## Installation

Copy the `custom_components/flood` folder into the config folder.

## Configuration

To add flood to your installation, go to Configuration >> Integrations in the UI, click the button with + sign and from the list of integrations select Flood.

## Lovelace suggestion

![lovelace card](lovelace.jpg)

```yaml
type: vertical-stack # or custom:vertical-stack-in-card if you have it
cards:
  - cards:
      - type: markdown
        content: >-
          <center><a href="http://IP_FLOOD" target="_blank"><img
          src="https://flood.js.org/assets/img/logo.png"
          width="60"></a></center>
      - type: button
        name: Stop
        show_state: false
        tap_action:
          action: call-service
          service: flood.stop_all_torrents
          service_data: {}
          target: {}
        icon: mdi:download-off
        icon_height: 30px
      - type: button
        name: Start
        icon: mdi:download
        show_state: false
        tap_action:
          action: call-service
          service: flood.start_all_torrents
          service_data: {}
          target: {}
        icon_height: 30px
    type: grid
    columns: 3
    square: false
  - type: entities
    entities:
      - entity: sensor.flood_last_notification
        name: " "
  - type: horizontal-stack
    cards:
      - type: entities
        entities:
          - entity: select.flood_download_limit
          - entity: select.flood_upload_limit
      - type: entities
        entities:
          - entity: sensor.flood_torrents
          - entity: sensor.flood_current_download
            name: Download
          - entity: sensor.flood_current_upload
            name: Upload
```


CHANGELOG.md
# CHANGELOG

## 0.2.2

- Replace deprecated constant

## 0.2.1

- Fix auth error

## 0.2.0

- Replace deprecated method

## 0.1.1

- First release


custom_components/flood/entity.py
"""Support for the generic Flood entity."""

from homeassistant.helpers.update_coordinator import CoordinatorEntity

from .const import DOMAIN


class FloodEntity(CoordinatorEntity):
    """Representation of a Flood generic entity."""

    def __init__(
        self,
        controller,
        coordinator,
        name: str,
        category: str = None,
        key: str = None,
        icon: str = None,
        attributes: dict = None,
        max_speed_limit: int = 0,
    ):
        """Initialize the entity."""
        super().__init__(coordinator)
        self._controller = controller
        self._name = f"Flood {name}"
        self._category = category
        self._key = key
        self._icon = icon
        self._attributes = attributes
        self._max_speed_limit = max_speed_limit

    @property
    def device_info(self):
        """Return device information identifier."""
        return {
            "identifiers": {(DOMAIN, self._controller.host)},
            "via_device": (DOMAIN, self._controller.host),
        }

    @property
    def unique_id(self):
        """Return an unique id."""
        return "_".join(
            [
                DOMAIN,
                self._controller.host,
                self._name,
            ]
        )

    @property
    def name(self):
        """Return the name."""
        return self._name

    @property
    def icon(self):
        """Return the icon to use in the frontend."""
        return self._icon


custom_components/flood/select.py
"""Support for the Flood select."""

from homeassistant.components.select import SelectEntity
from homeassistant.const import UnitOfDataRate

from .const import CONTROLLER, COORDINATOR, DOMAIN
from .entity import FloodEntity


async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up the Flood platform."""
    data = hass.data[DOMAIN][config_entry.entry_id]
    cont = data[CONTROLLER]
    cdnt = data[COORDINATOR]

    entities = [
        FloodSpeedLimitEntity(
            cont,
            cdnt,
            "Download Limit",
            "client_settings",
            "throttleGlobalDownSpeed",
            "mdi:download-lock",
        ),
        FloodSpeedLimitEntity(
            cont,
            cdnt,
            "Upload Limit",
            "client_settings",
            "throttleGlobalUpSpeed",
            "mdi:upload-lock",
        ),
    ]

    async_add_entities(entities, True)


class FloodSpeedLimitEntity(FloodEntity, SelectEntity):
    """Representation of a Flood speed limit entity."""

    @property
    def unit_of_measurement(self) -> str:
        """Return the icon to use in the frontend."""
        return UnitOfDataRate.KILOBYTES_PER_SECOND

    @property
    def current_option(self):
        """Return the state."""
        byte_value = float(self.coordinator.data.get(self._category, {}).get(self._key))
        return str(round(byte_value / 1024))

    @property
    def options(self):
        """Return list of speed limit set in Flood settings."""
        byte_value = float(self.coordinator.data.get(self._category, {}).get(self._key))
        current = round(byte_value / 1024)

        options = []
        if self.coordinator.data.get("settings"):
            limits = self.coordinator.data["settings"].get("speedLimits", {})
            if self._key == "throttleGlobalDownSpeed":
                options = [round(number / 1024) for number in limits.get("download")]
            elif self._key == "throttleGlobalUpSpeed":
                options = [round(number / 1024) for number in limits.get("upload")]

        if current not in options:
            options.append(current)

        options.sort()

        return [str(i) for i in options]

    async def async_select_option(self, option: str) -> None:
        """Update the current value."""
        speed = int(option)
        if self._key == "throttleGlobalDownSpeed":
            await self._controller.set_download_limit(speed)
        elif self._key == "throttleGlobalUpSpeed":
            await self._controller.set_upload_limit(speed)
        await self.coordinator.async_request_refresh()


custom_components/flood/sensor.py
"""Support for the Flood sensors."""
import logging

from homeassistant.const import UnitOfDataRate

from .const import CONTROLLER, COORDINATOR, DOMAIN
from .entity import FloodEntity

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up the Flood platform."""
    data = hass.data[DOMAIN][config_entry.entry_id]
    cont = data[CONTROLLER]
    cdnt = data[COORDINATOR]

    entities = [
        FloodSpeedSensorEntity(
            cont,
            cdnt,
            "Current Download",
            "history",
            "downloadSpeed",
            "mdi:download",
        ),
        FloodSpeedSensorEntity(
            cont,
            cdnt,
            "Current Upload",
            "history",
            "uploadSpeed",
            "mdi:upload",
        ),
        FloodSensorEntity(
            cont,
            cdnt,
            "Last notification",
            "last_notification",
            "title",
            "mdi:comment-outline",
            attributes=["type", "torrent"],
        ),
        FloodSensorEntity(
            cont,
            cdnt,
            "Torrents",
            "torrents",
            "count",
            "mdi:file",
            attributes=["seeding", "downloading", "completed", "active", "inactive"],
        ),
    ]

    async_add_entities(entities, True)


class FloodSpeedSensorEntity(FloodEntity):
    """Representation of a Flood sensor."""

    @property
    def unit_of_measurement(self) -> str:
        """Return the icon to use in the frontend."""
        return UnitOfDataRate.KILOBYTES_PER_SECOND

    @property
    def state(self) -> int:
        """Return the state."""
        byte_value = float(self.coordinator.data.get(self._category, {}).get(self._key))
        return int(byte_value / 1024)

    @property
    def state_attributes(self):
        """Return the state attributes."""
        if self._attributes and self.coordinator.data.get(self._category, {}):
            attributes = {}
            for attribute in self._attributes:
                attributes.update(
                    {
                        attribute: self.coordinator.data.get(self._category, {}).get(
                            attribute
                        )
                    }
                )
            return attributes


class FloodSensorEntity(FloodEntity):
    """Representation of a Flood sensor."""

    @property
    def state(self):
        """Return the state."""
        if self.coordinator.data.get(self._category):
            return self.coordinator.data[self._category].get(self._key)

    @property
    def state_attributes(self):
        """Return the state attributes."""
        if self._attributes and self.coordinator.data.get(self._category, {}):
            attributes = {}
            for attribute in self._attributes:
                attributes.update(
                    {
                        attribute: self.coordinator.data.get(self._category, {}).get(
                            attribute
                        )
                    }
                )
            return attributes


custom_components/flood/config_flow.py
"""Config flow to configure the Flood integration."""
from aiohttp import CookieJar
import voluptuous as vol

from homeassistant import config_entries
from homeassistant.const import (
    CONF_HOST,
    CONF_PASSWORD,
    CONF_PORT,
    CONF_SCAN_INTERVAL,
    CONF_USERNAME,
)
from homeassistant.helpers.aiohttp_client import async_create_clientsession

from .const import DEFAULT_SCAN_INTERVAL, DOMAIN
from .pyflood import FloodApi, FloodCannotConnectError, FloodInvalidAuthError

BASE_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_HOST): str,
        vol.Required(CONF_PORT, default=80): int,
        vol.Required(CONF_SCAN_INTERVAL, default=DEFAULT_SCAN_INTERVAL): int,
        vol.Optional(CONF_USERNAME): str,
        vol.Optional(CONF_PASSWORD): str,
    }
)


@config_entries.HANDLERS.register(DOMAIN)
class FloodConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
    """Handle a Flood config flow."""

    VERSION = 1
    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL

    async def async_step_user(self, user_input=None):
        """Handle a flow initialized by the user."""
        errors = {}
        if user_input is None:
            return self.async_show_form(
                step_id="user", data_schema=BASE_SCHEMA, errors=errors
            )

        entry = await self.async_set_unique_id(f"{DOMAIN}, {user_input.get(CONF_HOST)}")

        if entry:
            self._abort_if_unique_id_configured()

        session = async_create_clientsession(
            self.hass, verify_ssl=False, cookie_jar=CookieJar(unsafe=True)
        )

        controller = FloodApi(
            user_input.get(CONF_HOST),
            user_input.get(CONF_PORT),
            user_input.get(CONF_USERNAME),
            user_input.get(CONF_PASSWORD),
            session=session,
        )

        try:
            await controller.auth()
            await controller.connected
        except FloodInvalidAuthError:
            errors["base"] = "invalid_auth"
            return self.async_show_form(
                step_id="user", data_schema=BASE_SCHEMA, errors=errors
            )
        except FloodCannotConnectError:
            errors["base"] = "cannot_connect"
            return self.async_show_form(
                step_id="user", data_schema=BASE_SCHEMA, errors=errors
            )

        return self.async_create_entry(title=user_input.get(CONF_HOST), data=user_input)


custom_components/flood/binary_sensor.py
"""Support for the Flood binary_sensors."""
import logging

from .const import CONTROLLER, COORDINATOR, DOMAIN
from .entity import FloodEntity

_LOGGER = logging.getLogger(__name__)


async def async_setup_entry(hass, config_entry, async_add_entities):
    """Set up the Flood platform."""
    data = hass.data[DOMAIN][config_entry.entry_id]
    cont = data[CONTROLLER]
    cdnt = data[COORDINATOR]

    entities = [
        FloodConnectedEntity(cont, cdnt, "Backend connection", "connected", "status"),
    ]

    async_add_entities(entities, True)


class FloodConnectedEntity(FloodEntity):
    """Representation of a Flood sensor."""

    @property
    def device_class(self):
        """Return the icon to use in the frontend."""
        return "connectivity"

    @property
    def state(self):
        """Return the state."""
        if self.coordinator.data.get(self._category):
            return self.coordinator.data[self._category].get(self._key)


custom_components/flood/pyflood/__init__.py
"""Get information from Flood."""
import asyncio
from operator import itemgetter
import socket

import aiohttp
import async_timeout


class FloodApi:
    """Class representing the Flood and its API."""

    def __init__(
        self,
        host: str,
        port: int = 80,
        username: str = None,
        password: str = None,
        request_timeout: int = 10,
        session: aiohttp.client.ClientSession = None,
    ) -> None:
        """Init a Flood API."""
        self._host = host
        self._port = port
        self._username = username
        self._password = password
        self._request_timeout = request_timeout
        self._api_url = f"http://{host}:{port}/api/"
        self._version = None
        self._mac_address = None

        self._session = session
        self._close_session = False

    async def _request(
        self, url: str, method: str, content: dict = None, params: dict = None
    ) -> dict:
        """Make a request to get data."""
        try_count = 0
        while True:
            try_count += 1
            if self._session is None:
                jar = aiohttp.CookieJar(unsafe=True)
                self._session = aiohttp.ClientSession(
                    headers={"Content-Type": "application/json"}, cookie_jar=jar
                )
                self._close_session = True

            try:
                with async_timeout.timeout(self._request_timeout):
                    response = await self._session.request(
                        method=method, url=url, json=content, params=params
                    )
            except asyncio.TimeoutError as exception:
                raise FloodCannotConnectError(
                    "Timeout occurred while connecting to Flood."
                ) from exception
            except (aiohttp.ClientError, socket.gaierror) as exception:
                raise FloodCannotConnectError(
                    "Error occurred while communicating with Flood."
                ) from exception
            if response.status == 401:
                if try_count < 3:
                    self.auth()
                    continue
                raise FloodInvalidAuthError("Authentication failed with Flood.")

            if response.status:  # == 200:
                result = await response.json()
                response.close()
                return result
            else:
                raise FloodCannotConnectError("Unknown error")

    @property
    def host(self) -> str:
        """Return host name or ip used."""
        return self._host

    @property
    async def connected(self) -> bool:
        """Check if flood is connected to torrent client."""
        data = await self._request(
            method="GET", url=self._api_url + "client/connection-test"
        )
        return data.get("isConnected") is True

    @property
    async def history(self) -> dict:
        """Get all client settings."""
        history = await self._request(
            method="GET",
            url=self._api_url + "history",
            params={"snapshot": "FIVE_MINUTE"},
        )
        return {
            "downloadSpeed": history.get("download")[-1]
            if history.get("download")
            else 0,
            "uploadSpeed": history.get("upload")[-1] if history.get("upload") else 0,
        }

    async def auth(self) -> None:
        """Get authentication status after send credentials."""
        data = await self._request(
            method="POST",
            url=self._api_url + "auth/authenticate",
            content={"username": self._username, "password": self._password},
        )
        if "success" not in data:
            raise FloodInvalidAuthError()

    async def global_get(self) -> dict:
        """Get all data."""
        return {
            "client_settings": await self._request(
                method="GET", url=self._api_url + "client/settings"
            ),
            "settings": await self._request(
                method="GET", url=self._api_url + "settings"
            ),
            "last_notification": await self.last_notification(),
            "history": await self.history,
            "torrents": await self.torrents(),
            "connected": {"status": await self.connected},
        }

    async def client_settings(self) -> dict:
        """Get all client settings."""
        return await self._request(method="GET", url=self._api_url + "client/settings")

    async def notifications(self) -> dict:
        """Get all notifications."""
        return await self._request(method="GET", url=self._api_url + "notifications")

    async def last_notification(self):
        """Get last notifications."""
        NOTIF = "notifications"
        api_notifications = await self._request(method="GET", url=self._api_url + NOTIF)
        if NOTIF not in api_notifications or not api_notifications[NOTIF]:
            return {"title": "No notification"}

        last_notification = api_notifications[NOTIF][0]

        torrent_name = last_notification.get("data").get(
            "name", last_notification.get("data").get("title", "torrent name not found")
        )

        type_notification = last_notification.get("id", "notification type not found")
        if type_notification == "notification.torrent.finished":
            type_notification = "Finished"
        elif type_notification == "notification.torrent.errored":
            type_notification = "Errored"
        elif type_notification == "notification.feed.torrent.added":
            feed_name = last_notification.get("data").get(
                "feedLabel", "feed name not found"
            )
            type_notification = f"Added from {feed_name}"

        return {
            "title": f"{type_notification}: {torrent_name}",
            "type": last_notification.get("id", "notification type not found"),
            "torrent": torrent_name,
        }

    async def torrents(self) -> dict:
        """Get all client settings."""
        api_torrents = await self._request(method="GET", url=self._api_url + "torrents")
        torrents = api_torrents["torrents"].values()
        completed = list(filter(lambda d: "complete" in d["status"], torrents))
        seeding = list(filter(lambda d: "seeding" in d["status"], torrents))
        downloading = list(filter(lambda d: "downloading" in d["status"], torrents))
        active = list(filter(lambda d: "active" in d["status"], torrents))
        inactive = list(filter(lambda d: "inactive" in d["status"], torrents))
        stopped = list(filter(lambda d: "stopped" in d["status"], torrents))

        return {
            "count": len(torrents),
            "completed": len(completed),
            "downloading": len(downloading),
            "seeding": len(seeding),
            "stopped": len(stopped),
            "inactive": len(inactive),
            "active": len(active),
        }

    async def set_download_limit(self, speed: int) -> None:
        """Set download speed limit in kB/s."""
        await self._request(
            method="PATCH",
            url=self._api_url + "client/settings",
            content={"throttleGlobalDownSpeed": speed * 1024},
        )

    async def set_upload_limit(self, speed: int) -> None:
        """Set upload speed limit in kB/s."""
        await self._request(
            method="PATCH",
            url=self._api_url + "client/settings",
            content={"throttleGlobalUpSpeed": speed * 1024},
        )

    async def stop_all(self) -> None:
        """Stop all torrents."""
        api_torrents = await self._request(method="GET", url=self._api_url + "torrents")
        torrents = api_torrents["torrents"].values()
        torrents_hashes = {"hashes": list(map(itemgetter("hash"), torrents))}
        await self._request(
            method="POST",
            url=self._api_url + "torrents/stop",
            content=torrents_hashes,
        )

    async def start_all(self) -> None:
        """Start all torrents."""
        api_torrents = await self._request(method="GET", url=self._api_url + "torrents")
        torrents = api_torrents.get("torrents", {}).values()
        torrents_hashes = {"hashes": list(map(itemgetter("hash"), torrents))}
        await self._request(
            method="POST",
            url=self._api_url + "torrents/start",
            content=torrents_hashes,
        )

    async def close(self) -> None:
        """Close open client session."""
        if self._session and self._close_session:
            await self._session.close()

    async def __aenter__(self):
        """Async enter."""
        return self

    async def __aexit__(self, *_exc_info) -> None:
        """Async exit."""
        await self.close()


class FloodCannotConnectError(Exception):
    """Exception to indicate an error in connection."""


class FloodInvalidAuthError(Exception):
    """Exception to indicate an error in authentication."""
